### DISCUSSION TITLE
Mathics3-Rubi repository created! Rubi discussion moved.

### DISCUSSION THREAD

**Discussion by rocky** [Mathics3/mathics-core#1178] [Announcements]
Thanks to @aravindh-krishnamoorthy, serious progress has been made in getting Rubi to work on Mathics3.

To facilitate working on Rubi for Mathics3, [a new repository under the Mathics3 organization](https://github.com/Mathics3/Mathics3-Rubi) has been created. 

The discussion that had been started here in "core" has been moved to [there](https://github.com/Mathics3/Mathics3-Rubi/discussions). 

If you have questions, concerns, or comments about this, let me know.

**Comment by stblake** [comment]
This is good news! 

I am the defacto maintainer of Rubi after Albert's passing. I am also developing a C implementation of a tiny mathematica clone - picocas (https://github.com/stblake/picocas) which has a pattern matcher that can support most Rubi rules.

**Reply by aravindh-krishnamoorthy** [reply]
Yes, Sam. Albert's passing was a big loss to Rubi. Is your focus now to develop the if-then-else version that fits with Python, or are you going to continue to focus on the current pattern matching version? I will start working on Mathics3-Rubi again soon with the intention of making Mathics3 the go-to place for Rubi (schools, universities, etc.). In fact, I use Mathics3 quite a bit already for day-to-day research and would suggest that to you too :) I'll also take a peek into picocas... thanks!

**Reply by rocky** [reply]
> I am the defacto maintainer of Rubi after Albert's passing.

I am sorry to learn of Albert Rich's passing. I teleconferenced with him once. I believe he was living in Hawaii and was retired at that time. He seemed like a nice guy.

As it says in [Rubi's About](https://rulebasedintegration.org/about.html)

> Albert Rich
> 
> In 1978, David Stoutemyer and Albert co-founded The Soft Warehouse to develop and sell mathematical software for what eventually became known as “personal computers”. Albert was the author of the 8080, 8086 and Pentium assembly language implementations of muLisp, An Artificial Intelligence Development System the company sold. He and David were the co-authors of the company’s muMath (1979) and later Derive (1988) computer algebra systems. Theresa Shelby wrote the Visual C++ interface for Derive for Windows (1992). The company incorporated as Soft Warehouse, Inc. in 1983, and was sold, along with rights to its software, to Texas Instruments Inc. in 1998.

As for picocas and the other Mathematica implementations, SymJa, Woxi, and ?...

One thing I'd like to see is a more organized way of sharing information and implementation ideas.

Part of the reason for the Mathematica [named characters and operator information YAML tables](https://github.com/Mathics3/Mathics3-scanner/tree/master/mathics_scanner/data) was to facilitate use in other Mathematica implementations and projects. (I believe Symja has a copy of some of this information somewhere)

And there is @TiagoCavalcante's [mathics-threejs-backend](https://www.npmjs.com/package/@mathicsorg/mathics-threejs-backend) which is used not only in [Mathics3-django](https://github.com/Mathics3/Mathics3-django), but also in @axkr 's [Symja](https://github.com/axkr/symja_android_library).

And of course, I am sure one implementation is getting ideas and code from other open-source implementations. I know that I do.

In the section of the Mathics3-development guide, [Mathics3-Kernel-Interpreter](https://mathics-development-guide.readthedocs.io/en/latest/code-overview.html), I lay out how Mathics3's interpreter works, not just so that others might understand our code, but also so that other implementors might be able to use the ideas at a higher level.

But it would be good to have some sort of forum for discussing matters of _implementing_ Mathematica, either outside of Mathics3 or inside.

If someone creates such a forum or consortium, I will join.

One thing I note between all of the current implementations of Mathematica is that they all work off of an S-expression tree (or technically, an M-expression tree). Often this also goes under the name AST (Abstract Syntax Tree).

However, for performance, compilers (as opposed to interpreters) typically work off of _instructions_. In Mathics3, we've started moving the object-oriented evaluation methods out of the OO system and into a more function-oriented and instruction-friendly functions under the `mathics.eval` module.

This is the kind of stuff, i.e., strategies for implementing Mathematica, that I see this more general forum for.

Another idea that has come up recently is Bruce Lucas' @bdlucas1 idea of working with vectors via numpy and Mathematica's `NumericArray` (and `SparseArray`?). The idea here is that we just declare these kinds of objects are not the full generalized kind of numbers that individual numbers that Mathematica supports. And for plotting, which is what he is specifically interested in, this distinction does not matter.

As for picocas in particular, I tried to build it, and it looks to me like the file `configure.ac` was not added to the Git repository. See [Issue 3](https://github.com/stblake/picocas/issues/3).

**Comment by axkr** [comment]
Some notes from my side:

Maybe a bit "old school" in these AI days, but open source implementations can be inserted here:
- https://mathematica.stackexchange.com/a/4456/21734

For a forum or discussion I would prefer Discord (but be aware that Discord has the same problems with fighting against spam as other forums/platforms):
- https://discord.gg/tYknzr2qam

For `Graphics` I've tried to switch to AI generated SVGGraphics:
- https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/graphics/SVGGraphics.java
- https://github.com/axkr/symja_android_library/tree/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/builtin/graphics

and for `Graphics3D` to switch to AI generated WebGL 
- https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/graphics/WebGLGraphics3D.java
- https://github.com/axkr/symja_android_library/tree/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/builtin/graphics3d

This project may be most promising for using as a frontend in the future, if there is an open source kernel which is really MMA compatible:
- https://github.com/WLJSTeam/wljs-notebook
